File: Unidad10/Ejercicio3/Domain/Entities/Persona.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Domain.Entities
{
    [Table("Personas")]
    public class Persona
    {
        #region ATRIBUTOS PRIVADOS
        private int _id;
        private string _nombre;
        private string _apellidos;
        private int _edad;
        private DateTime _fechaNacimiento;
        private string _direccion;
        private string _telefono;
        private int _idDepartamento;
        private string _foto;
        #endregion

        #region CONSTRUCTOR
        /// <summary>
        /// Constructor de la clase persona con todos sus atributos
        /// </summary>
        /// <param name="id"></param>
        /// <param name="nombre"></param>
        /// <param name="apellidos"></param>
        /// <param name="edad"></param>
        public Persona(int id, string nombre, string apellidos, int edad, DateTime fechaNacimiento, string direccion, string telefono, int idDepartamento, string foto)
        {
            _id = id;
            _nombre = nombre;
            _apellidos = apellidos;
            _edad = edad;
            _fechaNacimiento = fechaNacimiento;
            _direccion = direccion;
            _telefono = telefono;
            _idDepartamento = idDepartamento;
            _foto = foto;
        }
        /// <summary>
        /// Constructor de la clase persona vacío
        /// </summary>
        public Persona()
        {
            _id = 0;
            _nombre = "";
            _apellidos = "";
            _edad = 0;
            _direccion = "";
            _telefono = "";
            _idDepartamento = 0;
            _foto = "";
        }
        #endregion

        #region GETTERS Y SETTERS
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        [Column("ID")]
        public int ID
        {
            get
            {
                return _id;
            }
            set
            {
                _id = value;
            }
        }

        [Required]
        [StringLength(30)]
        [Column("Nombre")]
        public string Nombre
        {
            get
            {
                return _nombre;
            }
            set
            {
                _nombre = value;
            }
        }

        [Required]
        [StringLength(60)]
        [Column("Apellidos")]
        public string Apellidos
        {
            get
            {
                return _apellidos;
            }
            set
            {
                _apellidos = value;
            }
        }

        [NotMapped]
        public int Edad
        {
            get
            {
                return _edad;
            }
            set
            {
                _edad = value;
            }
        }

        [Column("FechaNacimiento")]
        public DateTime FechaNacimiento
        {
            get
            {
                return _fechaNacimiento;
            }
            set
            {
                _fechaNacimiento = value;
            }
        }

        [StringLength(60)]
        [Column("Direccion")]
        public string Direccion
        {
            get
            {
                return _direccion;
            }
            set
            {
                _direccion = value;
            }
        }

        [StringLength(15)]
        [Column("Telefono")]
        public string Telefono
        {
            get
            {
                return _telefono;
            }
            set
            {
                _telefono = value;
            }
        }

        [StringLength(255)]
        [Column("Foto")]
        public string Foto
        {
            get
            {
                return _foto;
            }
            set
            {
                _foto = value;
            }
        }

        [Required]
        [Column("IDDepartamento")]
        [ForeignKey("Departamento")]
        public int IDDepartamento
        {
            get
            {
                return _idDepartamento;
            }
            set
            {
                _idDepartamento = value;
            }
        }
        #endregion
    }
}

File: Unidad15/ListadoPersonasAngular/src/domain/entities/Persona.ts
// src/domain/entities/Persona.ts

export class Persona {
  private readonly _ID: number;
  private _Nombre: string;
  private _Apellidos: string;
  private _Telefono: string;
  private _Direccion?: string | null;
  private _Foto?: string | null;
  private _FechaNacimiento?: Date | null;
  private _IDDepartamento: number;

  constructor(
    ID: number,
    Nombre: string,
    Apellidos: string,
    Telefono: string,
    IDDepartamento: number,
    Direccion?: string | null,
    Foto?: string | null,
    FechaNacimiento?: Date | null
  ) {
    this._ID = ID;
    this._Nombre = Nombre;
    this._Apellidos = Apellidos;
    this._Telefono = Telefono;
    this._Direccion = Direccion ?? null; 
    this._Foto = Foto ?? null; 
    this._FechaNacimiento = FechaNacimiento ?? null;
    this._IDDepartamento = IDDepartamento;
  }

  get ID(): number {
    return this._ID;
  }

  get Nombre(): string {
    return this._Nombre;
  }

  set Nombre(value: string) {
    this._Nombre = value;
  }

  get Apellidos(): string {
    return this._Apellidos;
  }

  set Apellidos(value: string) {
    this._Apellidos = value;
  }

  get Telefono(): string {
    return this._Telefono;
  }

  set Telefono(value: string) {
    this._Telefono = value;
  }

  get Direccion(): string {
    return this._Direccion ?? "";
  }

  set Direccion(value: string) {
    this._Direccion = value;
  }

  get Foto(): string {
    return this._Foto ?? "";
  }

  set Foto(value: string) {
    this._Foto = value;
  }

  get FechaNacimiento(): Date | null {
    return this._FechaNacimiento ?? null;
  }

  set FechaNacimiento(value: Date) {
    this._FechaNacimiento = value;
  }

  get IDDepartamento(): number {
    return this._IDDepartamento;
  }

  set IDDepartamento(value: number) {
    this._IDDepartamento = value;
  }

  get Edad(): number {
    if (!this._FechaNacimiento) return 0;
    const hoy = new Date();
    let edad = hoy.getFullYear() - this._FechaNacimiento.getFullYear();
    const mes = hoy.getMonth() - this._FechaNacimiento.getMonth();
    if (mes < 0 || (mes === 0 && hoy.getDate() < this._FechaNacimiento.getDate())) {
      edad--;
    }
    return edad;
  }
}